p2p/src/net/libp2p/connectivity/types.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): A. Altonen
+
+//! Types and utils used by the `ConnectionManager`
+
+use crate::error::P2pError;
+use libp2p::{identify, Multiaddr, PeerId};
+
+#[derive(Debug, PartialEq, Eq)]
+/// Events used to command the swarm
+pub enum ControlEvent {
+    /// Close the connection to peer
+    CloseConnection { peer_id: PeerId },
+}
+
+#[derive(Debug)]
+/// Events used to report swarm behaviour to front-end
+pub enum BehaviourEvent {
+    /// Inbound connection accepted
+    InboundAccepted {
+        addr: Multiaddr,
+        peer_info: Box<identify::IdentifyInfo>,
+    },
+
+    /// Outbound connection accepted
+    OutboundAccepted {
+        addr: Multiaddr,
+        peer_info: Box<identify::IdentifyInfo>,
+    },
+
+    /// Connection closed
+    ConnectionClosed { peer_id: PeerId },
+
+    /// Connection error
+    ConnectionError { addr: Multiaddr, error: P2pError },
+}
+
+macro_rules! peer_info_equal {
+    ($a:ident, $b:ident) => {
+        $a.public_key == $b.public_key
+            && $a.protocol_version == $b.protocol_version
+            && $a.agent_version == $b.agent_version
+            && $a.listen_addrs == $b.listen_addrs
+            && $a.protocols == $b.protocols
+            && $a.observed_addr == $b.observed_addr
+    };
+}
+
+impl PartialEq for BehaviourEvent {
+    fn eq(&self, other: &Self) -> bool {
```

Comment:
Why implementing this manually and not just inheriting it?

Reply:
I can't inherit it because `identify::IdentifyInfo` doesn't implement it.

Reply:
Instead of manually implementing the eq for this whole thing, create a simple wrapper of IdentifyInfo in which you define equality, then use that instead, then you can inherit Eq. Usually defining equality like this is a landmine from my experience, given that this is the only way to do it in C++.

Reply:
I don't see how it's a landmine but I'll try adding a wrapper for `IdentifyInfo`

Reply:
The issue usually happens when you change the struct and forget to update the Eq impl. As a rule of thumb, any primitive that requires updating two or more places to keep the code consistent is a bad idea. That's why they invented the "spaceship operator" in C++, which solved part of the problem, but not the whole problem. Rust's inheritance solution is wonderful.